<a href="https://colab.research.google.com/github/22f1000003/Machine-Learnig-Practice/blob/week-1-machine-learning-prctice/satyam_Introduction_to__pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This is my**first experience** with the colab
the following is the list
* Item 1
* Item 2
   * subitem 1
   * subitem2

1.   List item
2.   List item
$$ v = w0 + w1×1$$  


In [363]:

import pandas as pd
pd.__version__



'2.2.2'

The primary data structure in pandas are implemented as two classes:
* DataFrame. which you can imagin as a relational data table with rows and named columns.
* Series. which is a single column. Each row can be labled via an index.
* A DataFrame contains one or more Series and a name for each series.
The data frame is a commonly used abstration for data manipulation. similar implimentations exist in Sparkand R.

In [364]:
from sklearn.datasets import load_diabetes
diabetes = load_diabetes( as_frame=True)
print(type(diabetes['data']))
print(diabetes.data_filename)


<class 'pandas.core.frame.DataFrame'>
diabetes_data_raw.csv.gz


In [365]:
?load_diabetes

In [366]:
df = diabetes['data']

In [367]:
print(diabetes.keys())

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename', 'data_module'])


In [368]:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

*Creating pandas dataframe from Series*

In [369]:
cities = pd.Series(['delhi','Bengalore', 'Chennai','kanpur'])
population = pd.Series([17000000, 12000000, 6000000,])



**A DataFrame in pandas is:

👉 like a dictionary of columns (Series), where each column is like an array**

In [370]:
city_info_df = pd.DataFrame({'City':cities , 'Population':population})


In [371]:
type(city_info_df)

pandas.core.frame.DataFrame

In [372]:
city_info_df

,City,Population
0,delhi,17000000.0
1,Bengalore,12000000.0
2,Chennai,6000000.0
3,kanpur,NaN


In [373]:
city_info_df.head()

,City,Population
0,delhi,17000000.0
1,Bengalore,12000000.0
2,Chennai,6000000.0
3,kanpur,NaN


* psutil (process and system utilities)
* is a Python library for retrieving information on running processes and system utilization (CPU, memory, disks, network, sensors, etc.) in a cross-platform way (Linux, Windows, macOS, BSD, Solaris).

In [374]:
import psutil
print(f"Cloud RAM used: {psutil.virtual_memory().percent}%")

Cloud RAM used: 13.9%


**Eploring Data in DataFrame**





Quikly examine few entries in the DataFrame. Say first five 5 and last 5.

In [375]:
df.head()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362
4,0.005383,-0.044642,-0.036385,0.021872,0.003935,0.015596,0.008142,-0.002592,-0.031988,-0.046641


In [376]:
df.tail()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
437,0.041708,0.050680,0.019662,0.059744,-0.005697,-0.002566,-0.028674,-0.002592,0.031193,0.007207
438,-0.005515,0.050680,-0.015906,-0.067642,0.049341,0.079165,-0.028674,0.034309,-0.018114,0.044485
439,0.041708,0.050680,-0.015906,0.017293,-0.037344,-0.013840,-0.024993,-0.011080,-0.046883,0.015491
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930
441,-0.045472,-0.044642,-0.073030,-0.081413,0.083740,0.027809,0.173816,-0.039493,-0.004222,0.003064


In [377]:
? df.head

In [378]:
df.head(n=5)

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362
4,0.005383,-0.044642,-0.036385,0.021872,0.003935,0.015596,0.008142,-0.002592,-0.031988,-0.046641


In [379]:
df.tail(n=3)

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
439,0.041708,0.050680,-0.015906,0.017293,-0.037344,-0.013840,-0.024993,-0.011080,-0.046883,0.015491
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930
441,-0.045472,-0.044642,-0.073030,-0.081413,0.083740,0.027809,0.173816,-0.039493,-0.004222,0.003064


Find out number of rows and column in the dataframe

In [380]:
df.shape

(442, 10)

* in this diabetea data i have 442 rows and 10  columns
* how to see the names of the columns

In [381]:
df.columns

Index(['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'], dtype='object')

In [382]:
list(df.columns)

['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6']

In [383]:
?load_diabetes


In [384]:
?DESCR

Object `DESCR` not found.


In [385]:
print(diabetes.DESCR)

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

**Lecture-3**

In [386]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 442 entries, 0 to 441
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   age     442 non-null    float64
 1   sex     442 non-null    float64
 2   bmi     442 non-null    float64
 3   bp      442 non-null    float64
 4   s1      442 non-null    float64
 5   s2      442 non-null    float64
 6   s3      442 non-null    float64
 7   s4      442 non-null    float64
 8   s5      442 non-null    float64
 9   s6      442 non-null    float64
dtypes: float64(10)
memory usage: 34.7 KB


let's get summary statistics on the dataframe

In [387]:
df.describe()

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
count,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02
mean,-2.511817e-19,1.230790e-17,-2.245564e-16,-4.797570e-17,-1.381499e-17,3.918434e-17,-5.777179e-18,-9.042540e-18,9.293722e-17,1.130318e-17
std,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02
min,-1.072256e-01,-4.464164e-02,-9.027530e-02,-1.123988e-01,-1.267807e-01,-1.156131e-01,-1.023071e-01,-7.639450e-02,-1.260971e-01,-1.377672e-01
25%,-3.729927e-02,-4.464164e-02,-3.422907e-02,-3.665608e-02,-3.424784e-02,-3.035840e-02,-3.511716e-02,-3.949338e-02,-3.324559e-02,-3.317903e-02
50%,5.383060e-03,-4.464164e-02,-7.283766e-03,-5.670422e-03,-4.320866e-03,-3.819065e-03,-6.584468e-03,-2.592262e-03,-1.947171e-03,-1.077698e-03
75%,3.807591e-02,5.068012e-02,3.124802e-02,3.564379e-02,2.835801e-02,2.984439e-02,2.931150e-02,3.430886e-02,3.243232e-02,2.791705e-02
max,1.107267e-01,5.068012e-02,1.705552e-01,1.320436e-01,1.539137e-01,1.987880e-01,1.811791e-01,1.852344e-01,1.335973e-01,1.356118e-01


In [388]:
df.describe?

In [389]:
df.describe(percentiles= [0.2,0.3,0.5,0.6,0.8])

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
count,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02,4.420000e+02
mean,-2.511817e-19,1.230790e-17,-2.245564e-16,-4.797570e-17,-1.381499e-17,3.918434e-17,-5.777179e-18,-9.042540e-18,9.293722e-17,1.130318e-17
std,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02,4.761905e-02
min,-1.072256e-01,-4.464164e-02,-9.027530e-02,-1.123988e-01,-1.267807e-01,-1.156131e-01,-1.023071e-01,-7.639450e-02,-1.260971e-01,-1.377672e-01
20%,-4.547248e-02,-4.464164e-02,-4.048038e-02,-4.009893e-02,-3.871969e-02,-3.695017e-02,-3.971921e-02,-3.949338e-02,-4.117617e-02,-3.835666e-02
30%,-2.730979e-02,-4.464164e-02,-2.991782e-02,-2.873752e-02,-2.633611e-02,-2.480001e-02,-2.867429e-02,-3.949338e-02,-2.796492e-02,-2.468771e-02
50%,5.383060e-03,-4.464164e-02,-7.283766e-03,-5.670422e-03,-4.320866e-03,-3.819065e-03,-6.584468e-03,-2.592262e-03,-1.947171e-03,-1.077698e-03
60%,1.628068e-02,5.068012e-02,5.218854e-03,8.100982e-03,8.062710e-03,8.706873e-03,8.142084e-03,-2.592262e-03,1.255119e-02,7.206516e-03
80%,4.170844e-02,5.068012e-02,4.229559e-02,4.941519e-02,3.943444e-02,3.952068e-02,3.759519e-02,3.430886e-02,3.885335e-02,3.620126e-02
max,1.107267e-01,5.068012e-02,1.705552e-01,1.320436e-01,1.539137e-01,1.987880e-01,1.811791e-01,1.852344e-01,1.335973e-01,1.356118e-01


here we transpose it by using T attribute so it is become more readeable

In [390]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,442.0,-2.511817e-19,0.047619,-0.107226,-0.037299,0.005383,0.038076,0.110727
sex,442.0,1.230790e-17,0.047619,-0.044642,-0.044642,-0.044642,0.050680,0.050680
bmi,442.0,-2.245564e-16,0.047619,-0.090275,-0.034229,-0.007284,0.031248,0.170555
bp,442.0,-4.797570e-17,0.047619,-0.112399,-0.036656,-0.005670,0.035644,0.132044
s1,442.0,-1.381499e-17,0.047619,-0.126781,-0.034248,-0.004321,0.028358,0.153914
s2,442.0,3.918434e-17,0.047619,-0.115613,-0.030358,-0.003819,0.029844,0.198788
s3,442.0,-5.777179e-18,0.047619,-0.102307,-0.035117,-0.006584,0.029312,0.181179
s4,442.0,-9.042540e-18,0.047619,-0.076395,-0.039493,-0.002592,0.034309,0.185234
s5,442.0,9.293722e-17,0.047619,-0.126097,-0.033246,-0.001947,0.032432,0.133597
s6,442.0,1.130318e-17,0.047619,-0.137767,-0.033179,-0.001078,0.027917,0.135612


**Selection of particular column s and its attribute and i can say here geting series**

In [391]:
from sklearn.datasets import load_diabetes
diabetes = load_diabetes(as_frame=True)
df = diabetes['data']

In [392]:
df.columns

Index(['age', 'sex', 'bmi', 'bp', 's1', 's2', 's3', 's4', 's5', 's6'], dtype='object')

In [393]:
df['age']

,age
0,0.038076
1,-0.001882
2,0.085299
3,-0.089063
4,0.005383
...,...
437,0.041708
438,-0.005515
439,0.041708
440,-0.045472


In [394]:
type?

In [395]:
type(df['age'])

pandas.core.series.Series

how to find particular value in particular series

In [396]:
df['age'][0]

np.float64(0.038075906433423026)

In [397]:
df['age'][:5]

,age
0,0.038076
1,-0.001882
2,0.085299
3,-0.089063
4,0.005383


In [398]:
df['age'][-5:]

,age
437,0.041708
438,-0.005515
439,0.041708
440,-0.045472
441,-0.045472


In [399]:
df['age'][100:200]

,age
100,0.016281
101,0.016281
102,-0.092695
103,0.059871
104,-0.027310
...,...
195,0.027178
196,-0.023677
197,0.048974
198,-0.052738


In [400]:
df[['age','sex']]

,age,sex
0,0.038076,0.050680
1,-0.001882,-0.044642
2,0.085299,0.050680
3,-0.089063,-0.044642
4,0.005383,-0.044642
...,...,...
437,0.041708,0.050680
438,-0.005515,0.050680
439,0.041708,0.050680
440,-0.045472,-0.044642


In [401]:
df[['age','sex']][:5]

,age,sex
0,0.038076,0.050680
1,-0.001882,-0.044642
2,0.085299,0.050680
3,-0.089063,-0.044642
4,0.005383,-0.044642


In [402]:
df[['age','sex']][-5:]

,age,sex
437,0.041708,0.050680
438,-0.005515,0.050680
439,0.041708,0.050680
440,-0.045472,-0.044642
441,-0.045472,-0.044642


* **USE df.loc[] and df.iloc to select only rows, only columns or both**



In [403]:
df.iloc[0] # INteger base location
# it gives first row in this dataframe

,0
age,0.038076
sex,0.050680
bmi,0.061696
bp,0.021872
s1,-0.044223
s2,-0.034821
s3,-0.043401
s4,-0.002592
s5,0.019907
s6,-0.017646


In [404]:
df.head(n=4)

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
0,0.038076,0.050680,0.061696,0.021872,-0.044223,-0.034821,-0.043401,-0.002592,0.019907,-0.017646
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
2,0.085299,0.050680,0.044451,-0.005670,-0.045599,-0.034194,-0.032356,-0.002592,0.002861,-0.025930
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362


In [405]:
df.loc[0 ,'age']  # Label-based selection
# here i can use

np.float64(0.038075906433423026)

in the loc specified lable of the row

In [406]:
df.loc[0,'sex']

np.float64(0.05068011873981862)

* iloc position of the data in the row note here only uses integer it is similar to access from array
and in the loc is similar to access to like dictionary  

In [407]:
df.iloc[0 , 2:7]
# first enteries is equal to row
# second enteries is equal to accessing column
# here i want zeroth row 2 to 6 number column data
# 2 = bmi , 3 = bp , 4 = s1, 5=s2, 6=s3

,0
bmi,0.061696
bp,0.021872
s1,-0.044223
s2,-0.034821
s3,-0.043401


In [408]:
df.loc[4:6,'age']

,age
4,0.005383
5,-0.092695
6,-0.045472


* Label-based selection just like ['age', 'sex']

In [409]:
df.loc[3,['age','sex']]

,3
age,-0.089063
sex,-0.044642


In [410]:
age_df_temp = df.loc[df.age<5.383060e-03]

In [411]:
age_df_temp

,age,sex,bmi,bp,s1,s2,s3,s4,s5,s6
1,-0.001882,-0.044642,-0.051474,-0.026328,-0.008449,-0.019163,0.074412,-0.039493,-0.068332,-0.092204
3,-0.089063,-0.044642,-0.011595,-0.036656,0.012191,0.024991,-0.036038,0.034309,0.022688,-0.009362
5,-0.092695,-0.044642,-0.040696,-0.019442,-0.068991,-0.079288,0.041277,-0.076395,-0.041176,-0.096346
6,-0.045472,0.050680,-0.047163,-0.015999,-0.040096,-0.024800,0.000779,-0.039493,-0.062917,-0.038357
9,-0.070900,-0.044642,0.039062,-0.033213,-0.012577,-0.034508,-0.024993,-0.002592,0.067737,-0.013504
...,...,...,...,...,...,...,...,...,...,...
435,-0.012780,-0.044642,-0.023451,-0.040099,-0.016704,0.004636,-0.017629,-0.002592,-0.038460,-0.038357
436,-0.056370,-0.044642,-0.074108,-0.050427,-0.024960,-0.047034,0.092820,-0.076395,-0.061176,-0.046641
438,-0.005515,0.050680,-0.015906,-0.067642,0.049341,0.079165,-0.028674,0.034309,-0.018114,0.044485
440,-0.045472,-0.044642,0.039062,0.001215,0.016318,0.015283,-0.028674,0.026560,0.044529,-0.025930


In [412]:
age_df_temp.iloc[0]


,1
age,-0.001882
sex,-0.044642
bmi,-0.051474
bp,-0.026328
s1,-0.008449
s2,-0.019163
s3,0.074412
s4,-0.039493
s5,-0.068332
s6,-0.092204


* below the code is give key error because in loc we put lable data which is in age_df_temp
it work like dict key but iloc is not give error if i put 2 index because it access as integer

In [413]:
age_df_temp.loc[2]
# 2 lable is not present in Age_df_temp

KeyError: 2

* that why i called loc Label-based selection

In [ ]:
age_df_temp.iloc[2]

**Lecture-4 Data selection using pandas | part 2**

In [428]:
import numpy as np

In [440]:
df = pd.DataFrame(np.random.rand(9,4), index=list("abcdefghi"),columns=list('ABCD'))

In [442]:
df.shape

(9, 4)

In [444]:
df.columns

Index(['A', 'B', 'C', 'D'], dtype='object')

In [446]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9 entries, a to i
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   A       9 non-null      float64
 1   B       9 non-null      float64
 2   C       9 non-null      float64
 3   D       9 non-null      float64
dtypes: float64(4)
memory usage: 360.0+ bytes


In [449]:
df.head()

,A,B,C,D
a,0.052993,0.604714,0.273691,0.303047
b,0.396825,0.040045,0.911077,0.076091
c,0.945359,0.917701,0.907338,0.124213
d,0.011831,0.862836,0.556812,0.853484
e,0.699750,0.391116,0.266088,0.236639


In [450]:
another_df = pd.DataFrame(
    np.random.rand(100,4),
    index=range(10,110),
    columns=list('abcd')
)

In [451]:
another_df.head()

,a,b,c,d
10,0.704359,0.276903,0.005624,0.867727
11,0.564213,0.182083,0.683357,0.130959
12,0.343485,0.793627,0.481469,0.740159
13,0.310995,0.385486,0.030157,0.000439
14,0.721230,0.309505,0.445518,0.423575


In [452]:
another_df.tail()

,a,b,c,d
105,0.642034,0.511542,0.752974,0.954852
106,0.677203,0.825852,0.811134,0.653683
107,0.289909,0.531511,0.420860,0.701799
108,0.820017,0.281375,0.820308,0.369060
109,0.927965,0.417551,0.379205,0.862171


In [454]:
df.loc[0]
# it throw error because zero is not present in index

KeyError: 0

In [455]:
df.index

Index(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i'], dtype='object')

In [456]:
#df.loc[0]
# because in df 0 is not present but a is present

In [457]:
df.loc['a']

,a
A,0.052993
B,0.604714
C,0.273691
D,0.303047


In [461]:
df.loc['a','A':'C']


,a
A,0.052993
B,0.604714
C,0.273691


In [462]:
df.loc['a', 'A':]

,a
A,0.052993
B,0.604714
C,0.273691
D,0.303047


In [466]:
df.iloc[0:4,:] # row is 0:4 and column is : (all)

,A,B,C,D
a,0.052993,0.604714,0.273691,0.303047
b,0.396825,0.040045,0.911077,0.076091
c,0.945359,0.917701,0.907338,0.124213
d,0.011831,0.862836,0.556812,0.853484


In [465]:
df.iloc[:,1]

,B
a,0.604714
b,0.040045
c,0.917701
d,0.862836
e,0.391116
f,0.604444
g,0.823739
h,0.660115
i,0.683482


lambda function which is use directly do not need to define function

In [473]:
selector = lambda df: df['A']>0

In [475]:
selector

<function __main__.<lambda>(df)>

In [474]:
df.loc[selector]

,A,B,C,D
a,0.052993,0.604714,0.273691,0.303047
b,0.396825,0.040045,0.911077,0.076091
c,0.945359,0.917701,0.907338,0.124213
d,0.011831,0.862836,0.556812,0.853484
e,0.699750,0.391116,0.266088,0.236639
f,0.417870,0.604444,0.820590,0.248901
g,0.927342,0.823739,0.077760,0.072818
h,0.711938,0.660115,0.231372,0.798911
i,0.941218,0.683482,0.411290,0.765943


In [476]:
selector = lambda df: df['A']>0.6

In [479]:
print(selector)

<function <lambda> at 0x78620a239d00>


In [478]:
df.loc[selector]

,A,B,C,D
c,0.945359,0.917701,0.907338,0.124213
e,0.699750,0.391116,0.266088,0.236639
g,0.927342,0.823739,0.077760,0.072818
h,0.711938,0.660115,0.231372,0.798911
i,0.941218,0.683482,0.411290,0.765943


In [494]:
selctor = lambda df: (df['A'] > 0.5) & (df['B']>0.2)

In [485]:
selctor

<function __main__.<lambda>(df)>

In [486]:
df.loc[selctor]

,A,B,C,D
c,0.945359,0.917701,0.907338,0.124213
g,0.927342,0.823739,0.077760,0.072818
i,0.941218,0.683482,0.411290,0.765943


In [497]:
condition_for_selector = (df['A']>0.5) & (df['B']>0.4)

In [498]:
df[condition_for_selector]

,A,B,C,D
c,0.945359,0.917701,0.907338,0.124213
g,0.927342,0.823739,0.077760,0.072818
h,0.711938,0.660115,0.231372,0.798911
i,0.941218,0.683482,0.411290,0.765943


In [499]:
condition_for_selector

,0
a,False
b,False
c,True
d,False
e,False
f,False
g,True
h,True
i,True


**week-1 LECTURE_5 : Drop in dataframe::**

**Adding a column inthe Dateframe**

Creating a new column E in my DataFrame df

np.random.rand.(9)
*   List item This comes from NumPy
*   List item 👉 It generates:

9 random numbers

Between 0 and 1



In [ ]:
df['E'] = np.random.rand(9)

In [ ]:
df.head()

In [ ]:
df['F'] = np.random.rand(9)*10

In [ ]:
df.head()

In [ ]:
df['G'] = df['E'] + df['F']

In [ ]:
df

In [ ]:
criteria = df['A']<0.2

In [ ]:
criteria

here you can see only one value is true

In [ ]:
df.loc[criteria,'A']

In [ ]:
df

In [ ]:
cities = ['satyam','aman','aditay','dsfds','dfsd','dfsf',
          'erwer','safas','sfaf']

In [ ]:
df['city'] = cities

In [ ]:
df

In [ ]:
criteria = df['city'].isin(['satyam','aman','sfaf'])

In [ ]:
criteria

In [ ]:
cities_new = ['satyam','aman','aditay','madhav','shuryans','dfsf',
          'erwer','safas','sfaf']

In [ ]:
df['new_city'] = cities_new

In [ ]:
df

In [ ]:
df_copy = df.copy()

In [ ]:
df

In [ ]:
df_copy

In [ ]:
criteria = df_copy['city'].isin(['satyam','aman','aditay'])

In [ ]:
criteria

**how to update the value the column**

In [ ]:
df_copy.loc[df.city == 'satyam', 'city'] = 'satyam_new'

In [ ]:
df_copy

In [ ]:
?df_copy.dorp

here we drop column

In [ ]:
df_copy.drop(['new_city'],axis=1)
# axis = 1 means do operation on the column and if axis = 0 it means operation on the row


In [ ]:
df_copy = df_copy.drop(['i'])

In [ ]:
df_copy

In [ ]:
?df_copy.sample

In [ ]:
df.sample(3)

In [ ]:
df_copy.sample(3)

random
-state it stop sample generation radomly after this it generate same sample

In [ ]:
df.sample(3,random_state=78)

In [ ]:
df.sample(3,random_state=78)

In [ ]:
df_copy.sample(3,replace = True)

**DataFrame.sample(n=None, frac=None, replace=False, weights=None, random_state=None, axis=None, ignore_index=False)**
these are following attribute of sample function
1. n👉 Number of rows to select
2. frac👉 Fraction (percentage) of data
3. replace👉 Allow repetition or not
False → no repeat (default)
True → repeat allowed
4. weights👉 Probability for each row
5. random_state👉 Fix randomness (for same result every time)
6. axis👉 Sample rows or columns
0 → rows (default)
1 → columns
ignore_index
👉 Reset index or not

**W1_L6: Apply statistics using pandas**

In [ ]:
import pandas as pd
cities = ['Mumbai','Chennai','Pune','Ahmedabad','Kolkata','Kanpur', 'Delhi']
city_df = pd.DataFrame(cities)

In [ ]:
city_df.columns = ['city_name']

In [ ]:
city_df

In [ ]:
condition = city_df.city_name == 'Mumbai'

In [ ]:
condition

In [ ]:
type(condition)


In [ ]:
city_df[condition]

In [ ]:
city_df[city_df.city_name == 'Pune']

**Aggregation and grouping**

In [ ]:
import numpy as np

In [414]:
random_state = np.random.RandomState(20)
random_series = pd.Series(random_state.randn(10))

In [415]:
random_series

,0
0,0.883893
1,0.195865
2,0.357537
3,-2.343262
4,-1.084833
5,0.559696
6,0.939469
7,-0.978481
8,0.503097
9,0.406414


**Now that the series is created, we can try aggregation function like mean,
std etc.**

In [416]:
random_series.mean()

np.float64(-0.05606039362719746)

In [417]:
random_series.std()

1.0622481307589278

In [418]:
random_series.sum()

np.float64(-0.5606039362719746)

**let's created s dataframe with 5 rows adn 2 columns contaning random number**

In [419]:
df = pd.DataFrame({'A':random_state.rand(5), 'B':random_state.rand(5)})

In [420]:
df

,A,B
0,0.775245,0.254806
1,0.036664,0.857626
2,0.116694,0.949779
3,0.751281,0.561687
4,0.239218,0.178781


In [421]:
df.sum()

,0
A,1.919102
B,2.802678


In [422]:
df.mean()


,0
A,0.383820
B,0.560536


**Try these operation row_wise**

In [423]:
df.mean(axis=1)

,0
0,0.515025
1,0.447145
2,0.533236
3,0.656484
4,0.208999


In [424]:
df.sum(axis=1)

,0
0,1.030051
1,0.894290
2,1.066473
3,1.312968
4,0.417999


## Groupby
**Split **- we split datafram into smaller dataframe based on the value of keys
**Apply** - we apply desired
aggregation/trandsformation on each dataframe
**Combine**- we combine result from apply state into a datafram

In [425]:
df = pd.DataFrame({'key':['A','B','C']*2, #
                   'data': range(6)})

In [426]:
df

,key,data
0,A,0
1,B,1
2,C,2
3,A,3
4,B,4
5,C,5


In [427]:
df.groupby('key').sum()

,data
key,
A,3
B,5
C,7
